openwhisk/executor.go

```diff
@@ -0,0 +1,208 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package openwhisk
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"runtime"
+	"time"
+)
+
+// TIMEOUT to wait for process to start
+// and log to be produced
+const TIMEOUT = 5 * time.Millisecond
+
+// Executor is the container and the guardian  of a child process
+// It starts a command, feeds input and output, read logs and control its termination
+type Executor struct {
+	io      chan string
+	log     chan bool
+	exit    chan error
+	_cmd    *exec.Cmd
+	_input  *bufio.Writer
+	_output *bufio.Scanner
+	_logout *bufio.Scanner
+	_logerr *bufio.Scanner
+}
+
+// NewExecutor creates a child subprocess using the provided command line.
+// You can then start it getting a communication channel
+func NewExecutor(command string, args ...string) (proc *Executor) {
+	cmd := exec.Command(command, args...)
+
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return nil
+	}
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil
+	}
+
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil
+	}
+
+	pipeOut, pipeIn, err := os.Pipe()
+	if err != nil {
+		return nil
+	}
+	cmd.ExtraFiles = []*os.File{pipeIn}
+
+	return &Executor{
+		make(chan string),
+		make(chan bool),
+		make(chan error),
+		cmd,
+		bufio.NewWriter(stdin),
+		bufio.NewScanner(pipeOut),
+		bufio.NewScanner(stdout),
+		bufio.NewScanner(stderr),
+	}
+}
+
+// collect log from a stream
+func _collect(ch chan string, scan *bufio.Scanner) {
+	for scan.Scan() {
+		ch <- scan.Text()
+	}
+}
+
+// loop over the command executing
+// returning when the command exits
+func (proc *Executor) run() {
+	log.Println("run: start")
+	err := proc._cmd.Start()
+	if err != nil {
+		proc.exit <- err
+		log.Println("run: early exit")
+		proc._cmd = nil // do not kill
+		return
+	}
+	// wait for the exit
+	proc.exit <- proc._cmd.Wait()
+	proc._cmd = nil // do not kill
+	log.Println("run: end")
+}
+
+// manage copying stdout and stder in output
+// with log guards
+func (proc *Executor) logger() {
+	log.Println("logger: start")
+	// poll stdout and stderr
+	chOut := make(chan string)
+	go _collect(chOut, proc._logout)
+	chErr := make(chan string)
+	go _collect(chErr, proc._logerr)
+
+	// wait for the signal
+	for <-proc.log {
+		// flush stdout
+		runtime.Gosched()
+		for loop := true; loop; {
+			select {
+			case buf := <-chOut:
+				fmt.Println(buf)
+			case <-time.After(TIMEOUT):
+				loop = false
+			}
+		}
+		fmt.Println("XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX")
+
+		// flush stderr
+		runtime.Gosched()
+		for loop := true; loop; {
+			select {
+			case buf := <-chErr:
+				fmt.Println(buf)
+			case <-time.After(TIMEOUT):
+				loop = false
+			}
+		}
+		fmt.Println("XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX")
```

Comment:
is there a need for both `XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX` log guards .. maybe this is the way the protocol must work between the executor and proxy... ?

Reply:
It is not documented anywhere but it is how both the python proxy works

Reply:
For sure there is a need to know where the output of an action ends to be collected somO think it is required

Reply:
one marker per stream (stdout, stderr); this marker allows the log drain to insert an activation frame (when a single container is used for more than one activation.

Reply:
If the action runtime has `sentinelledLogs`, then the log reading code in the invoker looks for this string twice (once on stdout, once on stderr).   https://github.com/apache/incubator-openwhisk/blob/0b20df0f725a671f8e51c9e8793116476fd22f76/core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala#L51

Reply:
Indeed the go proxy drains the stderr and the stdout separately -I changed the protocol tu use channel 3 when realized how the guarded logs work

Reply:
Any change needed?

Reply:
ah.. sounds like no changes are needed, although if i understand this correctly, the stdout and stderr pipes are drained into a single pipe giving the combined output
```
... stdout stuff ....
XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX
... stderr stuff ...
XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX
```

Reply:
they are drained to docker stdout, exacly as docker the python proxy does.

Reply:
btw docker collect stderr and stdout in a single log all in stout, the reason I am copying both on stdout is because I add synchronization issues flushing them in separated streams

Reply:
:+1:

Reply:
you could i think collapse the two drain loops into a method... small nit; the only diff is `chErr` vs `chOut`.